### ISSUE TITLE
when reading journal file, it  lacks  condition judgment of  "object.size=0"

### ISSUE THREAD

**Issue by fayeinseu** [systemd/systemd#14943]
**systemd version the issue has been seen with**
> ...245 , 239

**Used distribution**
 > …239

**Expected behaviour you didn't see**
 > … when rsyslog calls sd_journal_next(), it runs well.

**Unexpected behaviour you saw**
 > …rsyslog sometimes opens deleted journal file and coredump。

**Steps to reproduce the problem**
1. a lot of logs flushes continuously.
2. journal files rotate or vacumm.
3. rsyslog calls sd_journal_next()，it reads an entry of size 0 .

**Comment by hexiaowen** [comment]
```
static void init_location(Location *l, LocationType type, JournalFile *f, Object *o) {
        assert(l);
        assert(IN_SET(type, LOCATION_DISCRETE, LOCATION_SEEK));
        assert(f);
        assert(o->object.type == OBJECT_ENTRY);    //**crash here, cause of o->object.type == 0**

        l->type = type;
        l->seqnum = le64toh(o->entry.seqnum);
```

and 

```
static int return_data(sd_journal *j, JournalFile *f, Object *o, const void **data, size_t *size) {
        size_t t;
        uint64_t l;
        int compression;

        l = le64toh(o->object.size) - offsetof(Object, data.payload);  // **goes wrong here, cause of o->object.size ==0** 
        t = (size_t) l;

        /* We can't read objects larger than 4G on a 32bit machine */
        if ((uint64_t) t != l)
```

We think things have gotten worse since the last journal_file_move_to_object, because at this time the journal unlinked and emptied the file to all 0s. journal files are stored in /run/log which is tmpfs.

**Comment by hexiaowen** [comment]
This will cause rsyslog crash when there are particularly many logs in the system.
Can anyone solve this problem？

**Comment by vcaputo** [comment]
I wonder if the immediate use of unlinkat_deallocate() when vacuuming is really the best solution, it certainly introduces some headaches like those illuminated by this issue.

Even if we manage to make all the reader code robust in the face of asynchronously zeroed journal data, is it really desirable to have journal data vanishing *during* legitimate access of it?  What if we just threw out some information relevant to whatever journal reading was in-progress, while performing a routine scheduled vacuum operation?

Maybe it makes sense to introduce some advisory file locking around vacuuming and reading, just to introduce a short wait before the unlinkat_deallocate() hammer when an exclusive lock is contended by readers holding shared locks.

**Comment by poettering** [comment]
So the reason we did the unlinkat_deallocate() was because unpriv clients could lock as much data onto disk otherwise as they wanted, simply by keeping the journal files open (for example by doing journalctl and then C-Z). That's why locking can't work either: we can't let journald block on unpriv clients...

**Comment by poettering** [comment]
My proposed fix is in #15557

**Comment by syyhao1994** [comment]
> My proposed fix is in #15557

I noticed that @fayeinseu make more judgements in function journal_file_check_object, real_journal_next and init_location. Why you drop these judgements in https://github.com/systemd/systemd/pull/15557? Did anywhere else do the judgement instead?
